### ISSUE TITLE
inverse predicates in RML

### ISSUE THREAD

**Issue by pmaria** [kg-construct/rml-fno-spec#40]
An aspect missing from RML is the ability to generate a triple using an inverse predicate.

Specifiying the generation of triples with an inverse predicate would allow you to generate a triple in the inverse direction so `object -> predicate -> subject` in the context of the subject generating triples map.

> Note: YARRRML [also introduces](https://rml.io/yarrrml/spec/#predicates-and-objects) inverse predicates.

Proposal:
- Add a new construct `rml:InversePredicateMap` which is a term map that can be referenced from a `rml:PredicateObjectMap` using `rml:inversePredicateMap`.
- Add a shortcut property `rml:inversePredicate` for `rml:inversePredicateMap`, analogous to `rr:predicate`.
- Redefine `rml:PredicateObjectMap` such that it should have at least one `rml:PredicateMap` or `rml:InversePredicateMap`. This will allow for maximum flexibility in use.
- Restrict the allowed term type of an rml:ObjectMap that is referenced by a `rml:PredicateObjectMap` which also rererences an `rml:InversePredicateMap` to IRIs and blank nodes.

Example:

```
<someTriplesMap>
  rml:logicalSource [] ;
  rr:subject ex:subject ;
  rr:predicateObjectMap [
    rr:predicateMap [
      rr:constant ex:parent ;
    ] ;
    rml:inversePredicateMap [
      rr:constant ex:child ;
    ] ;
    rr:object ex:object ;
  ] ;
```

or using shortcut properties:

```
<someTriplesMap>
  rml:logicalSource [] ;
  rr:subject ex:subject ;
  rr:predicateObjectMap [
    rr:predicate ex:parent ;
    rml:inversePredicate ex:child;
    rr:object ex:object ;
  ] ;
```

would generate:

```
ex:subject ex:parent ex:object .

ex:object ex:child ex:subject .
```

**Comment by andimou** [comment]
Overall I agree with this, but isn't it more a core issue rather than an fno issue?

Only a clarification, with the following

> Redefine rml:PredicateObjectMap such that it should have at least one rml:PredicateMap or rml:InversePredicateMap. This will allow for maximum flexibility in use.

you mean we can have a POM which only has an inverse predicate map but not a regular predicate map? or do you see more flexibility than this?

**Comment by pmaria** [comment]
> Overall I agree with this, but isn't it more a core issue rather than an fno issue?

Agreed, but we don't have a separate repository for core yet as far as I know.

> you mean we can have a POM which only has an inverse predicate map but not a regular predicate map? or do you see more flexibility than this?

Indeed, that is exactly what I meant.

**Comment by DylanVanAssche** [comment]
I like this! Implementation-wise this should be straightforward as well, no additional complexity.

**Comment by dachafra** [comment]
This is a very old issue and I don't know if we will implement or leave it for the Working Group. I'd like to hear opinions from @andimou @pmaria @DylanVanAssche @anaigmo

**Comment by DylanVanAssche** [comment]
I would like to see this, but depends on RML-Core to add it to the spec. Also needs test cases and SHACL shapes of course.

**Comment by pmaria** [comment]
We can leave it on working group, unless I have time to add it beforehand :)

**Comment by idomingu** [comment]
This feature would come handy. I support it, unless it adds extra complexity in the RML engine.
